File: ml-benchmark/synduce/list/atoi.f
Config NonLinear = true;
Config SampleSize = 6;
Config ExtraGrammar = "AutoLifter";

Inductive List = nil Unit | cons {Int, List};
Inductive CList = cnil Unit | single Int | concat {CList, CList};

base = 10;

spec = \xs: List. (fix (
  \f: List -> {Int, Int}. \x: List.
  match x with
    nil _ -> {0, 1}
  | cons {h, t} ->
    let result  = (f t) in
    {+ h (* 10 result.1), * 10 result.2}
  end
) xs).1;

cat = fix (
  \f: List -> List -> List. \a: List. \b: List.
  match a with
    nil _ -> b
  | cons {h, t} -> cons {h, (f t b)}
  end
);

repr = fix (
    \f: CList -> Compress List. \xs: CList.
    match xs with
      cnil _ -> nil unit
    | single x -> cons {x, nil unit}
    | concat {a, b} -> cat (f a) (f b)
    end
);

main = \x: CList. spec (repr x);

File: ml-benchmark/synduce/list/atoi.ml
config NonLinear = true
config SampleSize = 6
config EnableDeepCoder = true

type list =
  | Nil
  | Cons of int * list

type clist =
  | CNil
  | Single of int
  | Concat of clist * clist

let base = 10

let rec cat a b =
  match a with
  | Nil -> b
  | Cons (h, t) -> Cons (h, cat t b)

val repr: clist -> list compress
let rec repr xs =
  match xs with
  | CNil -> Nil
  | Single x -> Cons (x, Nil)
  | Concat (a, b) -> cat (repr a) (repr b)

let fi = function (x, _) -> x 
let se = function (_, x) -> x

let rec spec xs =
  let rec f x =
    match x with
    | Nil -> (0, 1)
    | Cons (h, t) ->
      let result = f t in
      (h + 10 * fi result, 10 * se result)
  in
  fi (f xs)

let prog x = spec (repr x)